### PR TITLE
fix: Changes CallManager to detect unhealthy state.

### PR DESCRIPTION
**Pull request by damencho**
Uses a dynamic pool of threads and if max number is reached we mark
jigasi as unhealthy and fail calls.

### CODE REVIEW

File: src/main/java/org/jitsi/jigasi/CallManager.java [hunk 1]
```diff
@@ -25,21 +25,83 @@
...
     private static final int POOL_SIZE = 5;
 
+    /**
+     * The thread pool to serve all call operations like answer and hangup.
+     * We initialize it with the 5 threads and can grow up to 4 time that size.
+     */
     private static ExecutorService threadPool
-        = Executors.newFixedThreadPool(POOL_SIZE);
+        = new ThreadPoolExecutor(
+                POOL_SIZE, POOL_SIZE * 4,
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jigasi/CallManager.java:46]: Why do we need minimum 5 threads? They don't do a lot of work, and 1 should suffice.

File: src/main/java/org/jitsi/jigasi/CallManager.java [hunk 2]
```diff
@@ -459,7 +518,15 @@ public synchronized static void hangupCall(Call   call,
         hangupCallThread.reasonCode = reasonCode;
         hangupCallThread.reason = reason;
 
-        threadPool.submit(hangupCallThread);
+        // if we are unhealthy, let's process error hangups
+        if (!healthy && reasonCode > 400)
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jigasi/CallManager.java:522]: Why the `reasonCode > 400` check? The only case where `healthy==false`, we suspect `submit()` will fail with an exception.

File: src/main/java/org/jitsi/jigasi/CallManager.java (hunk 1 above)
Comment (damencho) [reply src/main/java/org/jitsi/jigasi/CallManager.java:46]: Ok, I will update it.

File: src/main/java/org/jitsi/jigasi/CallManager.java (hunk 2 above)
Comment (damencho) [reply src/main/java/org/jitsi/jigasi/CallManager.java:522]: Ok, I will drop that check.

File: src/main/java/org/jitsi/jigasi/CallManager.java [hunk 3]
```diff
@@ -623,4 +690,60 @@ public static synchronized void restartPool()
 
         threadPool = Executors.newFixedThreadPool(POOL_SIZE);
     }
+
+    /**
+     * A thread factory that creates daemon threads.
+     *
+     * @author Pawel Domas
+     */
+    private static class DaemonThreadFactory
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jigasi/CallManager.java:699]: Is this a copy/paste from jicofo...?

Comment (bgrozev) [reply src/main/java/org/jitsi/jigasi/CallManager.java:699]: Please use `CustomizableThreadFactory` from jitsi-utils

File: src/main/java/org/jitsi/jigasi/CallManager.java [hunk 4]
```diff
@@ -623,4 +690,60 @@ public static synchronized void restartPool()
...
+                if (threadName != null)
+                {
+                    t.setName(threadName + "-" + t.getId());
+                }
+            }
+            return t;
+        }
+    }
+
+    private static class DummyCallThread
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jigasi/CallManager.java:729]: This is not used anywhere?

File: src/main/java/org/jitsi/jigasi/SipGatewaySession.java [hunk 5]
```diff
@@ -436,7 +445,16 @@ Exception onConferenceCallStarted(Call jvbConferenceCall)
...
         if (destination == null)
         {
-            CallManager.acceptCall(sipCall);
+            try
+            {
+                CallManager.acceptCall(sipCall);
+            }
+            catch(OperationFailedException e)
+            {
+                hangUpSipCall(e.getErrorCode(), "Cannot answer call");
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jigasi/SipGatewaySession.java:454]: Do we want to include the meessage as well (`e.getMessage()`)?

File: src/main/java/org/jitsi/jigasi/health/Health.java [hunk 6]
```diff
@@ -110,6 +110,12 @@ public static void stop()
     public static void check()
         throws Exception
     {
+        if (!CallManager.isHealthy())
+        {
+            throw new Exception(
+                "CallManager blocked to answer or hangup calls.");
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jigasi/health/Health.java:116]: I think just `CallManager is not healthy` should suffice as an explanation. `Health.java` shouldn't care about the specifics

File: src/main/java/org/jitsi/jigasi/CallManager.java (hunk 3 above)
Comment (damencho) [reply src/main/java/org/jitsi/jigasi/CallManager.java:708]: Yes.

File: src/main/java/org/jitsi/jigasi/CallManager.java (hunk 4 above)
Comment (damencho) [reply src/main/java/org/jitsi/jigasi/CallManager.java:738]: Yes. Just give a name of the threads for the stack traces.

File: src/main/java/org/jitsi/jigasi/SipGatewaySession.java (hunk 5 above)
Comment (damencho) [reply src/main/java/org/jitsi/jigasi/SipGatewaySession.java:454]: Well, that ends up in the sip message ... I was thinking it maybe too long ...

File: src/main/java/org/jitsi/jigasi/CallManager.java (hunk 3 above)
Comment (damencho) [reply src/main/java/org/jitsi/jigasi/CallManager.java:708]: Nice, thanks.

File: src/main/java/org/jitsi/jigasi/CallManager.java (hunk 4 above)
Comment (damencho) [reply src/main/java/org/jitsi/jigasi/CallManager.java]: Oh, this is something I forgot to delete :) I was filling pools with this one while testing :P

Comment (damencho) [reply src/main/java/org/jitsi/jigasi/CallManager.java]: It is deleted now.

**Comment by damencho** [comment]
Thanks.

**Pull request merged**